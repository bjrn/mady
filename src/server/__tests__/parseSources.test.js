/* eslint-env jest */
import {
  _getRegexps as getRegexps,
  _parseFunctionCalls as parseFunctionCalls,
  _parseReactIntl as parseReactIntl,
} from '../parseSources';

describe('Source parser', () => {
  describe('Function parser', () => {
    it('should work in a simple case', () => {
      const keys = {};
      const regexpFunctionNames = getRegexps(['FOO']);
      parseFunctionCalls(keys, 'FILENAME', `
        const a = 3;
        const str = FOO('Example');
      `, regexpFunctionNames);
      expect(keys).toMatchSnapshot();
    });

    it('should work with single and double quotes', () => {
      const keys = {};
      const regexpFunctionNames = getRegexps(['FOO']);
      parseFunctionCalls(keys, 'FILENAME', `
        const a = 3;
        const str = FOO('Example');
        const str2 = FOO("Example2");
      `, regexpFunctionNames);
      expect(keys).toMatchSnapshot();
    });

    it('should extract context if available', () => {
      const keys = {};
      const regexpFunctionNames = getRegexps(['FOO']);
      parseFunctionCalls(keys, 'FILENAME', `
        const a = 3;
        const str = FOO('context_Example');
      `, regexpFunctionNames);
      expect(keys).toMatchSnapshot();
    });

    it('should work with multiline strings', () => {
      const keys = {};
      const regexpFunctionNames = getRegexps(['FOO']);
      parseFunctionCalls(keys, 'FILENAME', `
        const a = 3;
        const str = FOO('context_This is a very
long
string');
      `, regexpFunctionNames);
      expect(keys).toMatchSnapshot();
    });
  });

  describe('React Intl parser', () => {
    it('should extract messages from FormattedMessage components', () => {
      const keys = {};
      parseReactIntl(keys, 'FILENAME', `
        import { FormattedMessage } from 'react-intl';
        const name = 'Guille';
        const Component = () => (
          <FormattedMessage
            id="reactIntlId1"
            defaultMessage={\`context_Hello {NAME}!\`}
            values={{ NAME: <b>{name}</b> }}
            description="Bla bla bla"
          />
        );
      `);
      expect(keys).toMatchSnapshot();
    });

    it('should extract messages from FormattedHTMLMessage components', () => {
      const keys = {};
      parseReactIntl(keys, 'FILENAME', `
        import { FormattedHTMLMessage } from 'react-intl';
        const name = 'Guille';
        const Component = () => (
          <FormattedHTMLMessage
            id="reactIntlId2"
            defaultMessage={\`context_<i>Hello {NAME}!</i>\`}
            values={{ NAME: name }}
          />
        );
      `);
      expect(keys).toMatchSnapshot();
    });

    it('should extract messages from defineMessages() calls', () => {
      const keys = {};
      parseReactIntl(keys, 'FILENAME', `
        import { defineMessages } from 'react-intl';
        defineMessages({
          reactIntlId3: {
            id: 'reactIntlId3',
            description: 'Message to greet the user',
            defaultMessage: 'someContext_Hello, {NAME}!',
          },
        });
      `);
      expect(keys).toMatchSnapshot();
    });
  });
});
