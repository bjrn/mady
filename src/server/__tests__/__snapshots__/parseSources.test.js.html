
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>Snapshot preview</title>
    <style>
      body.jest {
        display: flex;
        flex-direction: row;
        margin: 0;
        padding: 0;
        height: 100vh;
      }
      .jest--sidebar {
        width: 18em;
        padding: 0.3em 0;
        background-color: #eee;
        font-family: sans-serif;
        font-size: 0.8em;
      }
      .jest--preview {
        flex: 1;
        display: none;
        transform: translateZ(0);
      }
      .jest--preview.shown { display: block; }
      .jest--key {
        padding: 0.3em 1em;
        cursor: pointer;
      }
      .jest--key.shown { background-color: #ca461a; color: white; }
      .jest--key:hover { background-color: #ddd; }
      .jest--key.shown:hover { background-color: #ba360a; }
      pre.jest--nonreact { padding: 10px }
      
      
    </style>
    <script type="text/javascript">
      function removeClass(className, classToRemove) {
        var idx;
        var el;
        var classes;
        var els = document.getElementsByClassName(className);
        for (var i = 0; i < els.length; i++) {
          el = els[i];
          classes = el.className.split(' ');
          idx = classes.indexOf(classToRemove);
          if (idx >= 0) {
            classes.splice(idx, 1);
            el.className = classes.join(' ');
          }
        };
      }
      function onClick(id) {
        removeClass('jest--key', 'shown');
        removeClass('jest--preview', 'shown');
        el = document.getElementById('jest--preview_' + id);
        if (el) el.className += ' shown';
        el = document.getElementById('jest--key_' + id);
        if (el) el.className += ' shown';
      }
    </script>
  </head>
  <body class="jest">
    <div class="jest--sidebar">

      <div class="jest--key shown"
           id="jest--key_Source parser Custom regexp parsing should work with custom regexps 1"
           onclick="onClick('Source parser Custom regexp parsing should work with custom regexps 1')">
        Source parser Custom regexp parsing should work with custom regexps 1
      </div>

      <div class="jest--key"
           id="jest--key_Source parser Custom regexp parsing should work with custom regexps and multiline strings 1"
           onclick="onClick('Source parser Custom regexp parsing should work with custom regexps and multiline strings 1')">
        Source parser Custom regexp parsing should work with custom regexps and multiline strings 1
      </div>

      <div class="jest--key"
           id="jest--key_Source parser Function parser should extract context if available 1"
           onclick="onClick('Source parser Function parser should extract context if available 1')">
        Source parser Function parser should extract context if available 1
      </div>

      <div class="jest--key"
           id="jest--key_Source parser Function parser should work in a simple case 1"
           onclick="onClick('Source parser Function parser should work in a simple case 1')">
        Source parser Function parser should work in a simple case 1
      </div>

      <div class="jest--key"
           id="jest--key_Source parser Function parser should work with multiline strings 1"
           onclick="onClick('Source parser Function parser should work with multiline strings 1')">
        Source parser Function parser should work with multiline strings 1
      </div>

      <div class="jest--key"
           id="jest--key_Source parser Function parser should work with single and double quotes 1"
           onclick="onClick('Source parser Function parser should work with single and double quotes 1')">
        Source parser Function parser should work with single and double quotes 1
      </div>

      <div class="jest--key"
           id="jest--key_Source parser Function parser should work with specially-named functions 1"
           onclick="onClick('Source parser Function parser should work with specially-named functions 1')">
        Source parser Function parser should work with specially-named functions 1
      </div>

      <div class="jest--key"
           id="jest--key_Source parser React Intl parser should extract messages from FormattedHTMLMessage components 1"
           onclick="onClick('Source parser React Intl parser should extract messages from FormattedHTMLMessage components 1')">
        Source parser React Intl parser should extract messages from FormattedHTMLMessage components 1
      </div>

      <div class="jest--key"
           id="jest--key_Source parser React Intl parser should extract messages from FormattedMessage components 1"
           onclick="onClick('Source parser React Intl parser should extract messages from FormattedMessage components 1')">
        Source parser React Intl parser should extract messages from FormattedMessage components 1
      </div>

      <div class="jest--key"
           id="jest--key_Source parser React Intl parser should extract messages from defineMessages() calls 1"
           onclick="onClick('Source parser React Intl parser should extract messages from defineMessages() calls 1')">
        Source parser React Intl parser should extract messages from defineMessages() calls 1
      </div>

    </div>

    <div id="jest--preview_Source parser Custom regexp parsing should work with custom regexps 1" class="jest--preview shown">
      <pre class="jest--nonreact">
Object {
  "ISE=": Object {
    "context": null,
    "firstUsed": null,
    "id": "ISE=",
    "sources": Array [
      "FILENAME",
    ],
    "text": "!!",
    "unusedSince": null,
  },
  "aG8=": Object {
    "context": null,
    "firstUsed": null,
    "id": "aG8=",
    "sources": Array [
      "FILENAME",
    ],
    "text": "ho",
    "unusedSince": null,
  },
  "aGV5": Object {
    "context": null,
    "firstUsed": null,
    "id": "aGV5",
    "sources": Array [
      "FILENAME",
    ],
    "text": "hey",
    "unusedSince": null,
  },
  "bGV0J3MgZ28=": Object {
    "context": null,
    "firstUsed": null,
    "id": "bGV0J3MgZ28=",
    "sources": Array [
      "FILENAME",
    ],
    "text": "let\'s go",
    "unusedSince": null,
  },
}
</pre>
    </div>

    <div id="jest--preview_Source parser Custom regexp parsing should work with custom regexps and multiline strings 1" class="jest--preview">
      <pre class="jest--nonreact">
Object {}
</pre>
    </div>

    <div id="jest--preview_Source parser Function parser should extract context if available 1" class="jest--preview">
      <pre class="jest--nonreact">
Object {
  "Y29udGV4dF9FeGFtcGxl": Object {
    "context": "context",
    "firstUsed": null,
    "id": "Y29udGV4dF9FeGFtcGxl",
    "sources": Array [
      "FILENAME",
    ],
    "text": "Example",
    "unusedSince": null,
  },
}
</pre>
    </div>

    <div id="jest--preview_Source parser Function parser should work in a simple case 1" class="jest--preview">
      <pre class="jest--nonreact">
Object {
  "RXhhbXBsZQ==": Object {
    "context": null,
    "firstUsed": null,
    "id": "RXhhbXBsZQ==",
    "sources": Array [
      "FILENAME",
    ],
    "text": "Example",
    "unusedSince": null,
  },
}
</pre>
    </div>

    <div id="jest--preview_Source parser Function parser should work with multiline strings 1" class="jest--preview">
      <pre class="jest--nonreact">
Object {
  "Y29udGV4dF9UaGlzIGlzIGEgdmVyeQpsb25nCnN0cmluZw==": Object {
    "context": "context",
    "firstUsed": null,
    "id": "Y29udGV4dF9UaGlzIGlzIGEgdmVyeQpsb25nCnN0cmluZw==",
    "sources": Array [
      "FILENAME",
    ],
    "text": "This is a very
long
string",
    "unusedSince": null,
  },
}
</pre>
    </div>

    <div id="jest--preview_Source parser Function parser should work with single and double quotes 1" class="jest--preview">
      <pre class="jest--nonreact">
Object {
  "RXhhbXBsZQ==": Object {
    "context": null,
    "firstUsed": null,
    "id": "RXhhbXBsZQ==",
    "sources": Array [
      "FILENAME",
    ],
    "text": "Example",
    "unusedSince": null,
  },
  "RXhhbXBsZTI=": Object {
    "context": null,
    "firstUsed": null,
    "id": "RXhhbXBsZTI=",
    "sources": Array [
      "FILENAME",
    ],
    "text": "Example2",
    "unusedSince": null,
  },
}
</pre>
    </div>

    <div id="jest--preview_Source parser Function parser should work with specially-named functions 1" class="jest--preview">
      <pre class="jest--nonreact">
Object {
  "aG8=": Object {
    "context": null,
    "firstUsed": null,
    "id": "aG8=",
    "sources": Array [
      "FILENAME",
    ],
    "text": "ho",
    "unusedSince": null,
  },
  "aGk=": Object {
    "context": null,
    "firstUsed": null,
    "id": "aGk=",
    "sources": Array [
      "FILENAME",
    ],
    "text": "hi",
    "unusedSince": null,
  },
  "dGhlcmU=": Object {
    "context": null,
    "firstUsed": null,
    "id": "dGhlcmU=",
    "sources": Array [
      "FILENAME",
    ],
    "text": "there",
    "unusedSince": null,
  },
}
</pre>
    </div>

    <div id="jest--preview_Source parser React Intl parser should extract messages from FormattedHTMLMessage components 1" class="jest--preview">
      <pre class="jest--nonreact">
Object {
  "Y29udGV4dF88aT5IZWxsbyB7TkFNRX0hPC9pPg==": Object {
    "context": "context",
    "description": undefined,
    "firstUsed": null,
    "id": "Y29udGV4dF88aT5IZWxsbyB7TkFNRX0hPC9pPg==",
    "reactIntlId": "reactIntlId2",
    "sources": Array [
      "FILENAME",
    ],
    "text": "<i>Hello {NAME}!</i>",
    "unusedSince": null,
  },
}
</pre>
    </div>

    <div id="jest--preview_Source parser React Intl parser should extract messages from FormattedMessage components 1" class="jest--preview">
      <pre class="jest--nonreact">
Object {
  "Y29udGV4dF9IZWxsbyB7TkFNRX0h": Object {
    "context": "context",
    "description": "Bla bla bla",
    "firstUsed": null,
    "id": "Y29udGV4dF9IZWxsbyB7TkFNRX0h",
    "reactIntlId": "reactIntlId1",
    "sources": Array [
      "FILENAME",
    ],
    "text": "Hello {NAME}!",
    "unusedSince": null,
  },
}
</pre>
    </div>

    <div id="jest--preview_Source parser React Intl parser should extract messages from defineMessages() calls 1" class="jest--preview">
      <pre class="jest--nonreact">
Object {
  "c29tZUNvbnRleHRfSGVsbG8sIHtOQU1FfSE=": Object {
    "context": "someContext",
    "description": "Message to greet the user",
    "firstUsed": null,
    "id": "c29tZUNvbnRleHRfSGVsbG8sIHtOQU1FfSE=",
    "reactIntlId": "reactIntlId3",
    "sources": Array [
      "FILENAME",
    ],
    "text": "Hello, {NAME}!",
    "unusedSince": null,
  },
}
</pre>
    </div>

  </body>
</html>
